File: go/backend-http/03-multiple-countries.go
package main

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"log"
	"net/http"
)

var (
	apiKey  = "YOUR_API_KEY" // Hardcoded API key - replace with your actual key
	baseURL = "https://api.countriesdb.com"
)

func main() {

	// 3. Multiple country validation
	fmt.Println("Validating multiple country codes: US, CA, MX, FR")
	result := call("/api/validate/country", map[string]any{"code": []string{"US", "CA", "MX", "FR"}})

	var data map[string]any
	if err := json.Unmarshal([]byte(result), &data); err != nil {
		log.Fatalf("Failed to parse response: %v", err)
	}

	if results, ok := data["results"].([]any); ok {
		for _, item := range results {
			if itemMap, ok := item.(map[string]any); ok {
				if valid, ok := itemMap["valid"].(bool); !ok || !valid {
					log.Fatal("❌ Some countries are invalid")
				}
			}
		}
	}

	fmt.Println("✅ All countries are valid")
}

func call(path string, payload map[string]any) string {
	body, _ := json.Marshal(payload)
	req, err := http.NewRequest(http.MethodPost, baseURL+path, bytes.NewBuffer(body))
	if err != nil {
		log.Fatal(err)
	}

	req.Header.Set("Authorization", "Bearer "+apiKey)
	req.Header.Set("Content-Type", "application/json")

	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		log.Fatal(err)
	}
	defer resp.Body.Close()

	out, _ := io.ReadAll(resp.Body)
	if resp.StatusCode >= 400 {
		log.Fatalf("❌ Request failed (%d): %s", resp.StatusCode, out)
	}

	fmt.Println("Response:", string(out))
	return string(out)
}

File: ruby/backend-nethttp/03-multiple-countries.rb
require 'bundler/setup'
require 'net/http'
require 'json'
require 'uri'

# Set your CountriesDB private API key here
api_key = 'YOUR_API_KEY' # Replace with your actual API key
if api_key == 'YOUR_API_KEY'
  abort 'Please set your API_KEY in the script'
end

def call(api_key, path, payload)
  uri = URI("https://api.countriesdb.com#{path}")
  
  http = Net::HTTP.new(uri.host, uri.port)
  http.use_ssl = true
  
  request = Net::HTTP::Post.new(uri.path)
  request['Authorization'] = "Bearer #{api_key}"
  request['Content-Type'] = 'application/json'
  request.body = payload.to_json
  
  response = http.request(request)
  body = JSON.parse(response.body)
  raise "Request failed: #{response.code} #{body}" unless response.is_a?(Net::HTTPSuccess)
  
  body
end

# 3. Multiple country validation
begin
  puts 'Validating multiple country codes: US, CA, MX, FR'
  result = call(api_key, '/api/validate/country', { code: ['US', 'CA', 'MX', 'FR'] })
  puts "Response: #{JSON.pretty_generate(result)}"
  
  if result['results'] && result['results'].any? { |r| !r['valid'] }
    raise 'Some countries are invalid'
  end
  
  puts '✅ All countries are valid'
rescue => e
  puts "❌ Validation failed: #{e.message}"
  exit 1
end




